torch/csrc/cudnn/Conv.cpp

```diff
@@ -140,11 +179,48 @@ struct algorithm_search<cudnnConvolutionBwdDataAlgo_t> {
     return bwd_data_algos;
   }
 
-  static cudnnConvolutionBwdDataAlgoPerf_t findAlgorithm(cudnnHandle_t handle, const Convolution& conv) {
+  static cudnnConvolutionBwdDataAlgoPerf_t findAlgorithm(THCState* state,cudnnHandle_t handle, const Convolution& conv,
+							  void* in, void* out, void* wght) {
     int algoCount;
+    cudnnStatus_t err;
     cudnnConvolutionBwdDataAlgoPerf_t perfResults;
-    CHECK(cudnnFindConvolutionBackwardDataAlgorithm(handle, conv.wdesc.desc,
-        conv.odesc.desc, conv.cdesc.desc, conv.idesc.desc, 1, &algoCount, &perfResults));
+    cudnnConvolutionBwdDataAlgo_t algo[] = {
+         CUDNN_CONVOLUTION_BWD_DATA_ALGO_0,
+         CUDNN_CONVOLUTION_BWD_DATA_ALGO_1,
+         CUDNN_CONVOLUTION_BWD_DATA_ALGO_FFT,
+         CUDNN_CONVOLUTION_BWD_DATA_ALGO_FFT_TILING,
+         CUDNN_CONVOLUTION_BWD_DATA_ALGO_WINOGRAD,
+         CUDNN_CONVOLUTION_BWD_DATA_ALGO_WINOGRAD_NONFUSED
+    };
+    size_t max_ws_size = 0;
+    void *data = NULL; // workspace
+    
+    for(int i=0; i<6; i++) {
+        size_t sz;
+        err = cudnnGetConvolutionBackwardDataWorkspaceSize(handle, conv.wdesc.desc, conv.odesc.desc, conv.cdesc.desc, conv.idesc.desc, algo[i], &sz);
+        if(CUDNN_STATUS_SUCCESS != err || sz == 0 || sz < max_ws_size) continue;
+        if(NULL != data){
+	   THCudaFree(state, data);
+           data = NULL;
+	}
+	THCudaMalloc(state, &data, sz);
+	if(NULL != data){
+           max_ws_size = (sz<max_ws_size)? max_ws_size:sz;
+        }
+    }
+    if(NULL == data){ // in case we free mem before allocation of bigger chunk and failed on allocation after that
+       THCudaMalloc(state, &data, max_ws_size);
+       // if failed now then put workspace size equal to 0
+       max_ws_size = (NULL == data)? 0: max_ws_size;
+    }
```

Comment:
It seems that this loop over algos appears in all of these functions. It should be possible to only define algo arrays in here and extract this code into a function that's a template over the algo type.